src/NuGet.Clients/NuGet.SolutionRestoreManager/VulnerablePackagesInfoBar.cs

```diff
@@ -0,0 +1,145 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using NuGet.VisualStudio;
+using NuGet.VisualStudio.Telemetry;
+
+#nullable enable
+
+namespace NuGet.SolutionRestoreManager
+{
+    [Export(typeof(IVulnerabilitiesFoundService))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class VulnerablePackagesInfoBar : IVulnerabilitiesFoundService, IVsInfoBarUIEvents
+    {
+        IAsyncServiceProvider _asyncServiceProvider = AsyncServiceProvider.GlobalProvider;
+        private IVsInfoBarUIElement? _infoBarUIElement;
+        private bool _isVulnerablePackagesInfoBarVisible; // InfoBar is currently being displayed in the Solution Explorer
+        private bool _wasVulnerablePackagesInfoBarClosed = false; // InfoBar was closed by the user, using the 'x'(close) in the InfoBar
+        private bool _wasVulnerablePackagesInfoBarHide = false; // InfoBar was hid, this is caused because there are no more vulnerabilities to address
+        private uint? _eventCookie; // To hold the connection cookie
+
+        [Import]
+        private Lazy<IPackageManagerLaunchService>? PackageManagerUIStarter { get; set; }
+
+        public async Task HasAnyVulnerabilitiesInSolution(bool hasVulnerabilitiesInSolution, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // If the InfoBar was closed, don't show it for the rest of the VS session
+            // if the infobar is already visible, no work needed
+            if (_wasVulnerablePackagesInfoBarClosed || _isVulnerablePackagesInfoBarVisible)
+            {
+                return;
+            }
+
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            // Hide the InfoBar if Vulnerabilities were fixed
+            if (!hasVulnerabilitiesInSolution)
+            {
+                _infoBarUIElement?.Close(); // This is going to call OnClosed
+                _wasVulnerablePackagesInfoBarHide = false;
+                return;
+            }
+
+            // Initialize the InfoBar host in the SolutionExplorer window
+            IVsInfoBarHost? infoBarHost;
+            try
+            {
+                IVsUIShell? uiShell = await _asyncServiceProvider.GetServiceAsync<SVsUIShell, IVsUIShell>(throwOnFailure: false);
+                if (ErrorHandler.Failed(uiShell!.FindToolWindow((uint)__VSFINDTOOLWIN.FTW_fFindFirst, VSConstants.StandardToolWindows.SolutionExplorer, out var windowFrame)))
+                {
+                    NullReferenceException exception = new NullReferenceException(nameof(windowFrame));
```

Comment:
An NRE might be reasonable, but is it the most accurate description of the problem? Can we extract any failure state from the `FindToolWindow` return value?

Reply:
Looked at the documentation and only says that will throw an error code https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivsuishell.findtoolwindow?view=visualstudiosdk-2022 I don't think is more helpful. I looked at Index implementations of InfoBar and all most everyone doing this, do you have a suggestion on what can do better?

Reply:
I think we're better off throwing an Exception with a custom message. The exception can include the details such as the failure code. 
You can capture the failure code and then use `ErrorHandler.Failed` on that.